File: RevisionProject/python/1_ToDoList/ToDoList/todo_list_app.py
from typing import List


todo_list: List[str] = []


def add_todo() -> None:
    shall_exit_method = False
    while not shall_exit_method:
        description_input = input("Enter TODO description: ")
        if len(description_input) <= 0:
            print("The description cannot be empty.")
        elif description_input not in todo_list:
            todo_list.append(description_input)
            print(f'TODO successfully added: {description_input}')
            shall_exit_method = True
        else:
            print('The description must be unique.')


def show_all_todos() -> bool:
    if len(todo_list) > 0:
        index_count = 1
        for todo in todo_list:
            print(f'{index_count}. {todo}')
            index_count += 1
        return True
    else:
        print("No TODOs have been added yet.")
        return False


def remove_todo() -> None:
    todo_to_remove = input("Select the index of the TODO you want to remove: ")
    removal_index = int(todo_to_remove) - 1
    if todo_to_remove.isnumeric() and removal_index < len(todo_list):
        print(f"TODO removed: {todo_list.pop(removal_index)}")
    else:
        print("The given index is empty or not valid.")


if __name__ == "__main__":
    shall_exit: bool = False
    print("Hello")

    while not shall_exit:
        print("What do you want to do ?")
        print("[S]ee all TODOs")
        print("[A]dd a TODO")
        print("[R]emove a TODO")
        print("[E]xit")

        userInput = input("select one of the given S, A, R, E options: ")

        match userInput:

            case "A" | "a":
                add_todo()

            case "E" | "e":
                shall_exit: bool = True

            case "S" | "s":
                show_all_todos()

            case "R" | "r":
                if show_all_todos():
                    remove_todo()

            case _:
                print("Incorrect input")

        print()

File: RevisionProject/Csharp/1_ToDoList/TodoList/TodoListApp.cs
Console.WriteLine("Hello");
List<string> todoList = new();

bool shallExit = false;

while (!shallExit)
{
    Console.WriteLine("What do you want to do ?");
    Console.WriteLine("[S]ee all TODOs");
    Console.WriteLine("[A]dd a TODO");
    Console.WriteLine("[R]emove a TODO");
    Console.WriteLine("[E]xit");

    var userInput = Console.ReadLine();

    switch (userInput)
    {
        case "A":
        case "a":
            AddToTodo();
            break;

        case "E":
        case "e":
            shallExit = true;
            break;

        case "S":
        case "s":
            ShowAllTodo();
            break;

        case "R":
        case "r":
            if (ShowAllTodo())
            {
                RemoveToDo();
            }
            break;
    }

    Console.WriteLine();

}

bool ShowAllTodo()
{
    if (todoList.Count > 0)
    {
        var indexCount = 1;
        foreach (var todo in todoList)
        {
            ShowMessage($"{indexCount}. {todo}");
            indexCount++;
        }
        return true;
    }

    ShowMessage("No TODOs have been added yet.");
    return false;
}

void AddToTodo()
{
    bool shallExitFromMethod = false;

    while (!shallExitFromMethod)
    {
        Console.WriteLine("Enter the TODO description: ");
        var todoDescription = Console.ReadLine();

        if (todoDescription == null)
        {
            ShowMessage("The description cannot be empty.");
        }
        else if (!todoList.Contains(todoDescription))
        {
            todoList.Add(todoDescription);
            ShowMessage($"TODO sucessfully added: {todoDescription}");
            shallExitFromMethod = true;
        }
        else
        {
            ShowMessage("The description must be unique.");
        }
    }
}


void RemoveToDo()
{
    ShowMessage("Select the index of the TODO you want to remove: ");
    bool isParsed = int.TryParse(Console.ReadLine(), out int parsedInput);
    var index = parsedInput - 1;

    if (isParsed && index < todoList.Count)
    {
        var todoDescription = todoList[index];
        todoList.RemoveAt(index);
        ShowMessage($"TODO removed: {todoDescription}");
    }
    else
    {
        ShowMessage("The given index is empty or not valid .");
    }
}

void ShowMessage(string message)
{
    Console.WriteLine(message);
}